### DISCUSSION TITLE
Problem with getFirstListItem()

### DISCUSSION THREAD

**Discussion by dimitarnikolovv** [pocketbase/pocketbase#3546] [Q&A] [answered]
Hi, 

I am writing an app that allows the user to change their username. In order to check if the newly selected username already exists I wrote this bit of code:

```js
try {
      await locals.pb.collection('users').getFirstListItem(`username = "${data.username}"`);

      return message(
        updateUsernameForm,
        'Username is in use.',
        {status: 400}
      );

    } catch (err) {

      try {
        if (err.status === 404) {
          await locals.pb.collection('users').update(locals.user.id, { username: data.username });
          return message(updateUsernameForm, 'Success', {
            status: 200,
          });
        }
     ...
```

The problem is that the filter does not work properly and does return not found on existing usernames. Or I am doing something wrong.

**Comment by ganigeorgiev** [comment]
> The problem is that the filter does not work properly and does return not found on existing usernames.

`pb.collection('users').getFirstListItem('username = "missingUsername"')` throws 404 for me in case there is no user satisfying the provided filter (_tested with PocketBase v0.19.0 and JS SDK v0.18.3_).


In any case, you don't have to manually check for duplicated usernames as there is a builtin validator that will do that for you.

**Reply by dimitarnikolovv** [reply]
I figured it out already, but the behavior I got from the getFirstListItem doesn't seem right. It throws 404 if no user satisfies the filter but in my understanding, it should not throw 404 if there is a user satisfying the filter. I am only doing this check for the sake of more detailed error handling on the front end. Written like this it works, but it's not perfect in my case.


````js

try {
      await locals.pb.collection('users').update(locals.user.id, { username: data.username });
      return message(updateUsernameForm, 'Потребителското Ви име беше успешно променено', {
        status: 200,
      });
    } catch (err) {
      return message(
        updateUsernameForm,
        `<strong>${data.username}</strong> is not available`,
        {
          status: 400,
        }
      );
    }

````

**Reply by ganigeorgiev** [reply, marked as answer]
Are you sure that you have List API rule access for the user collection?
If no, then it is expected to throw an error even if a user matching the filter exist.

**Reply by ganigeorgiev** [reply]
>  Written like this it works, but it's not perfect in my case.

Why it is not perfect, what is the expected result?

You should be able to access the `username` field error by doing something like `err.response.data.username.message`.

**Reply by dimitarnikolovv** [reply]
Mainly because of l18n of the error messages. Anyway, the problem was my list rule which leads to another question. What should I check for, because I'm not entirely sure:

````
id = @request.auth.id || username = @request.query.filter.username
````

is the current list rule, but I have a doubt that the second part is not valid.

**Reply by ganigeorgiev** [reply]
> Mainly because of l18n of the error messages.

In this case you can make use of the `code` property of the error object, aka. `err.response.data.username.code`.

I don't understand your use case and I don't know what to recommend. The `filter` is a query parameter, meaning that its value is a plain string, aka. the entire `username="${data.username }"` expression.

**Reply by dimitarnikolovv** [reply]
I will probably use the error codes for reference and remove the getFirstListItem check entirely as it adds unnecessary complexity and API rules for sure. Thanks for the replies.
